File: frontend/src/lib/codegen.js
/**
 * Multi-language code generation from request state.
 *
 * Each generator takes a request object with:
 *   - method: string (GET, POST, PUT, PATCH, DELETE, etc.)
 *   - url: string (full URL)
 *   - headers: Array<{key: string, value: string}>
 *   - body: string (raw body content)
 *
 * And returns a formatted code snippet as a string.
 */

/** Escape a string for safe use inside double-quoted strings */
function esc(str) {
  return String(str).replace(/\\/g, "\\\\").replace(/"/g, '\\"');
}

/** Escape a single-quoted string */
function escSingle(str) {
  return String(str).replace(/\\/g, "\\\\").replace(/'/g, "\\'");
}

/** Escape a backtick template literal */
function escBacktick(str) {
  return String(str).replace(/\\/g, "\\\\").replace(/`/g, "\\`").replace(/\$/g, "\\$");
}

/** Filter out headers that shouldn't be explicitly set */
function shouldIncludeHeader(key) {
  const lower = key.toLowerCase();
  // Skip headers that are automatically set
  if (lower === "host" || lower === "content-length") return false;
  return true;
}

/** Get the content-type header value if present */
function getContentType(headers) {
  for (const h of headers) {
    if (h.key.toLowerCase() === "content-type") return h.value;
  }
  return null;
}

// ─── cURL ───────────────────────────────────────────────────────────────────

export function generateCurl({ method, url, headers, body }) {
  const lines = [`curl -X ${method.toUpperCase()}`];
  const ct = getContentType(headers);

  for (const { key, value } of headers) {
    if (!key.trim() || !shouldIncludeHeader(key)) continue;
    lines.push(`  -H "${esc(key)}: ${esc(value)}"`);
  }

  if (body && body.trim()) {
    const contentType = ct || "application/json";
    if (!headers.some((h) => h.key.toLowerCase() === "content-type")) {
      lines.push(`  -H "Content-Type: ${esc(contentType)}"`);
    }
    lines.push(`  -d '${escSingle(body.trim())}'`);
  }

  lines.push(`  "${esc(url)}"`);
  return lines.join(" \\\n");
}

// ─── JavaScript fetch ────────────────────────────────────────────────────────

export function generateFetch({ method, url, headers, body }) {
  const lines = [];
  const hasCustomHeaders = headers.some(
    ({ key }) => key.trim() && shouldIncludeHeader(key)
  );

  lines.push(
    `fetch("${esc(url)}", {`
  );
  lines.push(`  method: "${method.toUpperCase()}",`);

  if (hasCustomHeaders) {
    lines.push(`  headers: {`);
    for (const { key, value } of headers) {
      if (!key.trim() || !shouldIncludeHeader(key)) continue;
      lines.push(`    "${esc(key)}": "${esc(value)}",`);
    }
    lines.push(`  },`);
  }

  if (body && body.trim()) {
    lines.push(`  body: \`${escBacktick(body.trim())}\`,`);
  }

  lines.push(`});`);
  return lines.join("\n");
}

// ─── JavaScript axios ────────────────────────────────────────────────────────

export function generateAxios({ method, url, headers, body }) {
  const m = method.toLowerCase();
  const configProps = [];
  const hasCustomHeaders = headers.some(
    ({ key }) => key.trim() && shouldIncludeHeader(key)
  );

  if (hasCustomHeaders) {
    const hdrLines = [];
    hdrLines.push(`  headers: {`);
    for (const { key, value } of headers) {
      if (!key.trim() || !shouldIncludeHeader(key)) continue;
      hdrLines.push(`    "${esc(key)}": "${esc(value)}",`);
    }
    hdrLines.push(`  },`);
    configProps.push(hdrLines.join("\n"));
  }

  if (body && body.trim()) {
    configProps.push(`  data: \`${escBacktick(body.trim())}\`,`);
  }

  if (configProps.length > 0) {
    return [
      `axios.${m}("${esc(url)}", {`,
      configProps.join("\n"),
      `});`,
    ].join("\n");
  }

  return `axios.${m}("${esc(url)}");`;
}

// ─── Go net/http ─────────────────────────────────────────────────────────────

const goMethodMap = {
  POST: "http.MethodPost",
  PUT: "http.MethodPut",
  PATCH: "http.MethodPatch",
  DELETE: "http.MethodDelete",
  HEAD: "http.MethodHead",
  OPTIONS: "http.MethodOptions",
};

function goMethod(method) {
  return goMethodMap[method.toUpperCase()] || `"${method.toUpperCase()}"`;
}

export function generateGo({ method, url, headers, body }) {
  const lines = [];
  lines.push(`// Create request`);
  lines.push(
    body && body.trim()
      ? `body := strings.NewReader(\`${escBacktick(body.trim())}\`)`
      : `body := http.NoBody`
  );
  lines.push(
    `req, err := http.NewRequest(${goMethod(method)}, "${esc(url)}", body)`
  );
  lines.push(`if err != nil {`);
  lines.push(`    panic(err)`);
  lines.push(`}`);

  for (const { key, value } of headers) {
    if (!key.trim() || !shouldIncludeHeader(key)) continue;
    lines.push(`req.Header.Set("${esc(key)}", "${esc(value)}")`);
  }

  lines.push(``);
  lines.push(`// Send request`);
  lines.push(`client := &http.Client{}`);
  lines.push(`resp, err := client.Do(req)`);
  lines.push(`if err != nil {`);
  lines.push(`    panic(err)`);
  lines.push(`}`);
  lines.push(`defer resp.Body.Close()`);
  lines.push(``);
  lines.push(`// Read response`);
  lines.push(`data, err := io.ReadAll(resp.Body)`);
  lines.push(`if err != nil {`);
  lines.push(`    panic(err)`);
  lines.push(`}`);
  lines.push(`fmt.Println(string(data))`);

  return lines.join("\n");
}

// ─── Python requests ─────────────────────────────────────────────────────────

export function generatePython({ method, url, headers, body }) {
  const lines = [];
  lines.push(`import requests`);
  lines.push(``);

  let configLines = [];
  let configIndent = "";

  const hasCustomHeaders = headers.some(
    ({ key }) => key.trim() && shouldIncludeHeader(key)
  );

  if (hasCustomHeaders || (body && body.trim())) {
    configLines.push(`url = "${esc(url)}"`);
  }

  if (hasCustomHeaders) {
    configLines.push(`headers = {`);
    for (const { key, value } of headers) {
      if (!key.trim() || !shouldIncludeHeader(key)) continue;
      configLines.push(`    "${esc(key)}": "${esc(value)}",`);
    }
    configLines.push(`}`);
  }

  if (body && body.trim()) {
    configLines.push(`data = '''${body.trim()}'''`);
  }

  if (configLines.length > 0) {
    lines.push(...configLines);
    lines.push(``);
  }

  const args = [];
  if (hasCustomHeaders) args.push(`headers=headers`);
  if (body && body.trim()) args.push(`data=data`);

  if (args.length > 0 || configLines.length > 0) {
    const urlRef = configLines.length > 0 ? "url" : `"${esc(url)}"`;
    lines.push(`response = requests.${method.toLowerCase()}(${urlRef}${args.length ? ", " + args.join(", ") : ""})`);
  } else {
    lines.push(`response = requests.${method.toLowerCase()}("${esc(url)}")`);
  }

  lines.push(``);
  lines.push(`print(response.status_code)`);
  lines.push(`print(response.text)`);

  return lines.join("\n");
}

// ─── HTTPie ──────────────────────────────────────────────────────────────────

export function generateHTTPie({ method, url, headers, body }) {
  const m = method.toUpperCase();
  const showMethod = m !== "GET" && m !== "POST";
  const parts = ["http"];

  if (showMethod) {
    parts.push(m);
  }

  for (const { key, value } of headers) {
    if (!key.trim() || !shouldIncludeHeader(key)) continue;
    parts.push(`"${esc(key)}:${esc(value)}"`);
  }

  if (body && body.trim()) {
    // HTTPie infers JSON if the body looks like it
    parts.push(`--raw='${escSingle(body.trim())}'`);
  }

  parts.push(`"${esc(url)}"`);
  return parts.join(" ");
}

// ─── Registry ────────────────────────────────────────────────────────────────

/** Map of language key → { label, generator } */
export const LANGUAGES = [
  { key: "curl", label: "cURL", generate: generateCurl },
  { key: "fetch", label: "JavaScript — fetch", generate: generateFetch },
  { key: "axios", label: "JavaScript — axios", generate: generateAxios },
  { key: "go", label: "Go — net/http", generate: generateGo },
  { key: "python", label: "Python — requests", generate: generatePython },
  { key: "httpie", label: "HTTPie", generate: generateHTTPie },
];

/**
 * Generate code for a specific language.
 * Returns the code string or null if language not found.
 */
export function generateCode(langKey, request) {
  const lang = LANGUAGES.find((l) => l.key === langKey);
  if (!lang) return null;
  try {
    return lang.generate(request);
  } catch {
    return null;
  }
}

File: app/pkg/parser/codegen.go
package parser

import (
	"fmt"
	"strings"
)

// CodeLanguage identifies a target programming language for code generation.
type CodeLanguage string

// Supported code generation languages.
const (
	LangCurl   CodeLanguage = "curl"
	LangFetch  CodeLanguage = "fetch"
	LangAxios  CodeLanguage = "axios"
	LangGo     CodeLanguage = "go"
	LangPython CodeLanguage = "python"
	LangHTTPie CodeLanguage = "httpie"
)

// AllLanguages returns the list of all supported code generation languages.
func AllLanguages() []CodeLanguage {
	return []CodeLanguage{LangCurl, LangFetch, LangAxios, LangGo, LangPython, LangHTTPie}
}

// LanguageLabel returns a human-readable label for a code language.
func LanguageLabel(lang CodeLanguage) string {
	switch lang {
	case LangCurl:
		return "cURL"
	case LangFetch:
		return "JavaScript (fetch)"
	case LangAxios:
		return "JavaScript (axios)"
	case LangGo:
		return "Go (net/http)"
	case LangPython:
		return "Python (requests)"
	case LangHTTPie:
		return "HTTPie"
	default:
		return string(lang)
	}
}

// CodeGenRequest contains the request data needed for code generation.
type CodeGenRequest struct {
	Method  string
	URL     string
	Headers map[string]string
	Body    string
}

// GenerateCode produces a code snippet in the target language for the given request.
// Returns an error for unsupported languages.
func GenerateCode(req CodeGenRequest, lang CodeLanguage) (string, error) {
	// Use uppercase method for consistency
	req.Method = strings.ToUpper(req.Method)
	if req.Method == "" {
		req.Method = "GET"
	}

	switch lang {
	case LangCurl:
		return generateCurl(req), nil
	case LangFetch:
		return generateFetch(req), nil
	case LangAxios:
		return generateAxios(req), nil
	case LangGo:
		return generateGo(req), nil
	case LangPython:
		return generatePython(req), nil
	case LangHTTPie:
		return generateHTTPie(req), nil
	default:
		return "", fmt.Errorf("unsupported language: %s", lang)
	}
}

// escapeString escapes a string for safe inclusion in Go/JS/Python string literals.
func escapeString(s string) string {
	s = strings.ReplaceAll(s, "\\", "\\\\")
	s = strings.ReplaceAll(s, "\"", "\\\"")
	s = strings.ReplaceAll(s, "\n", "\\n")
	return s
}

// indentHeaderLines indents multi-line strings by the given prefix.
func indentHeaderLines(headers map[string]string, indent string) string {
	if len(headers) == 0 {
		return ""
	}
	var sb strings.Builder
	for k, v := range headers {
		sb.WriteString(indent)
		sb.WriteString(fmt.Sprintf("%q: %q,\n", k, v))
	}
	return sb.String()
}

// generateCurl produces a curl command.
func generateCurl(req CodeGenRequest) string {
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("curl -X %s", req.Method))

	if req.Method == "HEAD" {
		sb.WriteString(" -I")
	}

	// Headers
	for k, v := range req.Headers {
		sb.WriteString(fmt.Sprintf(" \\\n  -H %q", k+": "+v))
	}

	// Body (skip for GET/HEAD)
	if req.Body != "" && req.Method != "GET" && req.Method != "HEAD" {
		sb.WriteString(fmt.Sprintf(" \\\n  -d %q", req.Body))
	}

	sb.WriteString(fmt.Sprintf(" \\\n  %q", req.URL))
	return sb.String()
}

// generateFetch produces a JavaScript fetch snippet.
func generateFetch(req CodeGenRequest) string {
	var sb strings.Builder

	// Build options object
	sb.WriteString(fmt.Sprintf("fetch(%q, {\n", req.URL))
	sb.WriteString(fmt.Sprintf("  method: %q,\n", req.Method))

	// Headers
	if len(req.Headers) > 0 {
		sb.WriteString("  headers: {\n")
		for k, v := range req.Headers {
			sb.WriteString(fmt.Sprintf("    %q: %q,\n", k, v))
		}
		sb.WriteString("  },\n")
	}

	// Body
	if req.Body != "" && req.Method != "GET" && req.Method != "HEAD" {
		contentType := req.Headers["Content-Type"]
		isJSON := strings.Contains(contentType, "json")
		if isJSON {
			// Pretty-print JSON body if possible
			sb.WriteString(fmt.Sprintf("  body: JSON.stringify(%s),\n", formatJSObject(req.Body)))
		} else {
			sb.WriteString(fmt.Sprintf("  body: %q,\n", req.Body))
		}
	}

	sb.WriteString("})\n")
	sb.WriteString("  .then(response => response.json())\n")
	sb.WriteString("  .then(data => console.log(data))\n")
	sb.WriteString("  .catch(error => console.error('Error:', error));")
	return sb.String()
}

// formatJSObject attempts to format a JSON body as a JS object.
func formatJSObject(body string) string {
	// Try to find JSON-like structure and wrap it as an object
	trimmed := strings.TrimSpace(body)
	if strings.HasPrefix(trimmed, "{") || strings.HasPrefix(trimmed, "[") {
		return trimmed
	}
	// For non-JSON bodies, just quote it
	return fmt.Sprintf("%q", body)
}

// generateAxios produces a JavaScript axios snippet.
func generateAxios(req CodeGenRequest) string {
	var sb strings.Builder

	sb.WriteString("axios({\n")
	sb.WriteString(fmt.Sprintf("  method: %q,\n", strings.ToLower(req.Method)))
	sb.WriteString(fmt.Sprintf("  url: %q,\n", req.URL))

	// Headers
	if len(req.Headers) > 0 {
		sb.WriteString("  headers: {\n")
		for k, v := range req.Headers {
			sb.WriteString(fmt.Sprintf("    %q: %q,\n", k, v))
		}
		sb.WriteString("  },\n")
	}

	// Body
	if req.Body != "" && req.Method != "GET" && req.Method != "HEAD" {
		contentType := req.Headers["Content-Type"]
		isJSON := strings.Contains(contentType, "json")
		if isJSON {
			sb.WriteString(fmt.Sprintf("  data: %s,\n", formatJSObject(req.Body)))
		} else {
			sb.WriteString(fmt.Sprintf("  data: %q,\n", req.Body))
		}
	}

	sb.WriteString("})\n")
	sb.WriteString("  .then(response => console.log(response.data))\n")
	sb.WriteString("  .catch(error => console.error('Error:', error));")
	return sb.String()
}

// generateGo produces a Go net/http snippet.
func generateGo(req CodeGenRequest) string {
	var sb strings.Builder

	hasBody := req.Body != "" && req.Method != "GET" && req.Method != "HEAD"

	sb.WriteString("package main\n\n")
	sb.WriteString("import (\n")
	sb.WriteString("\t\"fmt\"\n")
	if hasBody {
		sb.WriteString("\t\"strings\"\n")
	}
	sb.WriteString("\t\"net/http\"\n")
	sb.WriteString("\t\"io\"\n")
	sb.WriteString(")\n\n")
	sb.WriteString("func main() {\n")

	if hasBody {
		contentType := req.Headers["Content-Type"]
		isJSON := strings.Contains(contentType, "json")
		if isJSON {
			sb.WriteString(fmt.Sprintf("\tbody := strings.NewReader(`%s`)\n", req.Body))
		} else {
			sb.WriteString(fmt.Sprintf("\tbody := strings.NewReader(%q)\n", req.Body))
		}
		sb.WriteString(fmt.Sprintf("\treq, err := http.NewRequest(%q, %q, body)\n", req.Method, req.URL))
	} else {
		sb.WriteString(fmt.Sprintf("\treq, err := http.NewRequest(%q, %q, nil)\n", req.Method, req.URL))
	}
	sb.WriteString("\tif err != nil {\n\t\tpanic(err)\n\t}\n\n")

	// Headers
	for k, v := range req.Headers {
		sb.WriteString(fmt.Sprintf("\treq.Header.Set(%q, %q)\n", k, v))
	}

	sb.WriteString("\n")
	sb.WriteString("\tresp, err := http.DefaultClient.Do(req)\n")
	sb.WriteString("\tif err != nil {\n\t\tpanic(err)\n\t}\n")
	sb.WriteString("\tdefer resp.Body.Close()\n\n")
	sb.WriteString("\tbodyBytes, err := io.ReadAll(resp.Body)\n")
	sb.WriteString("\tif err != nil {\n\t\tpanic(err)\n\t}\n\n")
	sb.WriteString("\tfmt.Println(resp.Status)\n")
	sb.WriteString("\tfmt.Println(string(bodyBytes))\n")
	sb.WriteString("}")
	return sb.String()
}

// generatePython produces a Python requests snippet.
func generatePython(req CodeGenRequest) string {
	var sb strings.Builder

	hasBody := req.Body != "" && req.Method != "GET" && req.Method != "HEAD"

	sb.WriteString("import requests\n\n")

	if len(req.Headers) > 0 {
		sb.WriteString("headers = {\n")
		for k, v := range req.Headers {
			sb.WriteString(fmt.Sprintf("    %q: %q,\n", k, v))
		}
		sb.WriteString("}\n\n")
	}

	methodLower := strings.ToLower(req.Method)

	if hasBody {
		contentType := req.Headers["Content-Type"]
		isJSON := strings.Contains(contentType, "json")
		if isJSON {
			sb.WriteString(fmt.Sprintf("data = %s\n\n", formatPyObject(req.Body)))
		} else {
			sb.WriteString(fmt.Sprintf("data = %q\n\n", req.Body))
		}
		sb.WriteString(fmt.Sprintf("response = requests.%s(%q, headers=headers, data=data)\n", methodLower, req.URL))
	} else if len(req.Headers) > 0 {
		sb.WriteString(fmt.Sprintf("response = requests.%s(%q, headers=headers)\n", methodLower, req.URL))
	} else {
		sb.WriteString(fmt.Sprintf("response = requests.%s(%q)\n", methodLower, req.URL))
	}

	sb.WriteString("\nprint(response.status_code)\n")
	sb.WriteString("print(response.json())")
	return sb.String()
}

// formatPyObject attempts to format a JSON body as a Python dict/list.
func formatPyObject(body string) string {
	trimmed := strings.TrimSpace(body)
	if strings.HasPrefix(trimmed, "{") || strings.HasPrefix(trimmed, "[") {
		// Replace JSON null with None, true/false with True/False
		s := strings.ReplaceAll(trimmed, "null", "None")
		s = strings.ReplaceAll(s, "true", "True")
		s = strings.ReplaceAll(s, "false", "False")
		return s
	}
	return fmt.Sprintf("%q", body)
}

// generateHTTPie produces an HTTPie command.
func generateHTTPie(req CodeGenRequest) string {
	var sb strings.Builder

	method := strings.ToUpper(req.Method)
	if method == "GET" {
		// HTTPie defaults to GET
		sb.WriteString("http ")
	} else {
		sb.WriteString(fmt.Sprintf("http %s ", method))
	}

	sb.WriteString(req.URL)

	// Headers
	for k, v := range req.Headers {
		sb.WriteString(fmt.Sprintf(" \\\n  %s:%s", k, v))
	}

	// Body
	if req.Body != "" && method != "GET" && method != "HEAD" {
		contentType := req.Headers["Content-Type"]
		isJSON := strings.Contains(contentType, "json")
		if isJSON {
			sb.WriteString(fmt.Sprintf(" \\\n  %s", req.Body))
		} else {
			sb.WriteString(fmt.Sprintf(" \\\n  %s", req.Body))
		}
	}

	return sb.String()
}
